### PR TITLE
Add proposal for breaking changes

### PR DESCRIPTION
**Pull request by richlander**
A few of us have been thinking about how to better handle breaking changes in this repo. I've taken those ideas and written them in the doc. Please share your feedback.

If this new model is successful, we'll apply it to other repos.

Conversation started here: https://github.com/dotnet/runtime/issues/37689

Several other people helped on this proposal: @AaronRobinsonMSFT, @jkotas, @marklio, @PriyaPurkayastha, @jamshedd, @stephentoub, @danmosemsft

### CODE REVIEW

**Comment by jkotas** [comment]
Add link to this doc from https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/breaking-changes.md ?

File: docs/project/breaking-change-process.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+## Process
+
+1. Create an issue that describes the breaking change
+   * Mark the issue with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label
+   * Goals and motivation for the change
+   * Pre-change behavior
+   * Post-change behavior
+   * Errors or other behavior you can expect when running old code that breaks
+   * Workarounds and mitigations, including [AppContext switches](https://docs.microsoft.com/dotnet/api/system.appcontext)
```
Comment (danmosemsft) [inline docs/project/breaking-change-process.md:19]: Do we have guidance somewhere indicating when a change justifies adding a switch? My impression was that our bar was high for this, to avoid reducing maintainability. 

Currently I see a couple for globalization invariant mode, a couple for threadpool, 3 for HTTP/2, a couple others. So it seems we use them for 1-2% of breaking changes.

Also we can presumably say that the switch would always be to switch off the default behavior rather than vice versa.

**Review by danmoseley — APPROVED** [review]

File: docs/project/breaking-change-process.md (hunk 1 above)
Comment (richlander) [reply docs/project/breaking-change-process.md:19]: I am happy to remove this. I saw somewhat similar text in doc Priya gave me, and was wanting to pull some of thinking from it. I look to you for guidance on this. I'm fully neutral on it.

**Comment by richlander** [comment]
@jkotas -- it appears there is a link to that doc in the PR. Yes?

https://github.com/dotnet/runtime/pull/37779/files#diff-3e4368228ffed87f000e514b5d3fe593R42

**Comment by jkotas** [comment]
I meant link in the opposite direction to make this new doc easier to discover.

File: docs/project/breaking-change-process.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+You should start this process as soon as possible, preferably before any code is written, in the design phase. The breaking change issue should be paired with a design document that helps reviewers understand the broader scope, goals and constraints of your change. If you don't yet have a design document, you are encouraged to first write such a document, to enable people giving feedback to develop a deeper level of understanding on your approach. If you do not do this, you might stall the process.
+
+There are people on the team that can help you work through a breaking change, if you want to engage with them before starting this more formal process. They are part of the [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) . Feel free to @mention them on issues to start a dialogue.
+
+## Process
```
Comment (terrajobst) [inline docs/project/breaking-change-process.md:11]: Are we supposed to create new issues for issues that track design work in case this design has been identified as potentially being breaking or is it OK if we apply this label to such issues?

File: docs/project/breaking-change-process.md (hunk 1 above)
Comment (terrajobst) [reply docs/project/breaking-change-process.md:19]: I think instead of removing this I think it would be much better if your doc could start talking about what criteria we use to decide whether we do an app compat switch or not.

File: docs/project/breaking-change-process.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+   * Workarounds and mitigations, including [AppContext switches](https://docs.microsoft.com/dotnet/api/system.appcontext)
+   * Link to the issue for the feature or bug fix that the breaking change is associated with.
+   * Reference this issue from associated PRs.
+2. Share your issue with whomever you see as stakeholders
+   * Please @mention [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) team
+   * Engage with people commenting on the issue, with the goal of deriving the most feedback on your proposed breaking change
+   * This may involve a lot of explanation on your part. A well-written design doc can ease the burden here.
+3. Mark associated PRs with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label, and link to your breaking change issue.
+4. Once the PR is merged, create a [docs issue](https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md).
+   * Clarify which .NET preview the break will be first released in.
```
Comment (terrajobst) [inline docs/project/breaking-change-process.md:28]: Should we add an announcement in dotnet/announcement too? Presumably this makes sense for some, but not all. How do we decide?

File: docs/project/breaking-change-process.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+## Examples
+
+The following are good examples of breaking change issues
+
+* https://github.com/dotnet/runtime/issues/28788
+* https://github.com/dotnet/runtime/issues/37672
+
+
```
Comment (terrajobst) [inline docs/project/breaking-change-process.md:37]: nit: should probably be a single line

**Review by terrajobst — APPROVED** [review]
Left some questions, but I love this direction

File: docs/project/breaking-change-process.md (hunk 1 above)
Comment (jkotas) [reply docs/project/breaking-change-process.md:19]: >  talk about what criteria we use

Here is one: https://github.com/dotnet/runtime/issues/8655#issuecomment-415103199

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: docs/project/breaking-change-process.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Breaking change process
```
Comment (stephentoub) [inline docs/project/breaking-change-process.md:1]: Not all breaking changes are created equal.  Removing support for some feature is likely much more impactful than, say, switching the order of two argument validation exceptions that might occur if multiple invalid inputs are passed to a method.  Practically any bug fix can break someone.

What is the bar used for this process to define "breaking change"?

File: docs/project/breaking-change-process.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+Breaking changes make it difficult for people to adopt new versions of .NET. We should avoid making them. In the rare case that you believe a breaking change is unavoidable, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen  opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.
```
Comment (stephentoub) [inline docs/project/breaking-change-process.md:3]:
```suggestion
Breaking changes make it difficult for people to adopt new versions of .NET. We strive to avoid making them. In the rare case when a breaking change is unavoidable or the best option when all factors are considered, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen  opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.
```

File: docs/project/breaking-change-process.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+This process does not result in approval or denial of a breaking change request. It results in feedback, which might be varied or uniform. In the best case, the feedback is a call to action to you, that provides a direction on how you should proceed. The people that are responsible for reviewing your PR are expected to use the information collected by this process as part of their determination of whether your change is appropriate to merge.
+
+You should start this process as soon as possible, preferably before any code is written, in the design phase. The breaking change issue should be paired with a design document that helps reviewers understand the broader scope, goals and constraints of your change. If you don't yet have a design document, you are encouraged to first write such a document, to enable people giving feedback to develop a deeper level of understanding on your approach. If you do not do this, you might stall the process.
```
Comment (stephentoub) [inline docs/project/breaking-change-process.md:7]: > You should start this process as soon as possible, preferably before any code is written, in the design phase.

For smaller efforts, those are often one in the same, e.g. you don't discover that there needs to be a change in behavior until you're almost done implementing whatever it is you're implementing.

Comment (stephentoub) [inline docs/project/breaking-change-process.md:7]: > The breaking change issue should be paired with a design document that helps reviewers understand the broader scope, goals and constraints of your change.

Wouldn't the issue itself just contain all of those details?  Why is a separate "document" needed?

File: docs/project/breaking-change-process.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+There are people on the team that can help you work through a breaking change, if you want to engage with them before starting this more formal process. They are part of the [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) . Feel free to @mention them on issues to start a dialogue.
+
+## Process
+
+1. Create an issue that describes the breaking change
```
Comment (stephentoub) [inline docs/project/breaking-change-process.md:13]: A common workflow is there's an issue opened to describe a feature, the feature is discussed, and then subsequently a PR is opened to implement that feature.

Is this process proposing yet another issue be opened to describe potential breaks in the first issue?

File: docs/project/breaking-change-process.md (hunk 3 above)
Comment (richlander) [reply docs/project/breaking-change-process.md:28]: I think this would be a rare case.

File: docs/project/breaking-change-process.md (hunk 5 above)
Comment (richlander) [reply docs/project/breaking-change-process.md:1]: I have a few thoughts on this:

- It's hard to define a definite bar. 
- The breaking change docs that I linked to were the ones that were in the repo, mostly from the old corefx repo, I believe. We can file issues if we think they are inadequate.
- In terms of determination, I think of it as a layered approach. People should make a first determination in the early design phase. It would be late, but the person pressing the "big green button" needs to make a determination of the information available on whether a change is breaking. The last resort is .NET users coming to this conclusion in preview or stable releases.

In short, this document doesn't define that, and I think that's fine.

**Comment by richlander** [comment]
I think I covered the feedback, now. Thanks!

Good?

File: docs/project/breaking-change-process.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+This process does not result in approval or denial of a breaking change request. It results in feedback, which might be varied or uniform. In the best case, the feedback is a call to action to you, that provides a direction on how you should proceed. The people that are responsible for reviewing your PR are expected to use the information collected by this process as part of their determination of whether your change is appropriate to merge.
+
+You should start this process as soon as possible, preferably before any code is written, in the design phase. This isn't always possible, since the change in behavior might not be discovered until you're almost done building a feature. In the case of a significant breaking change, you  should provide reviewers with a design document so that they can understand the broader scope, goals and constraints of your change.
```
Comment (jkotas) [inline docs/project/breaking-change-process.md:7]:
```suggestion
You should start this process as soon as possible, preferably before any code is written, in the design phase. This isn't always possible, since the change in behavior might not be discovered until you're almost done building a feature. In the case of a significant breaking change, you should provide reviewers with a design document so that they can understand the broader scope, goals and constraints of your change.
```
(extra space)

File: docs/project/breaking-change-process.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
+# Breaking change process
+
+Breaking changes make it difficult for people to adopt new versions of .NET. We strive to avoid making them. In the rare case when a breaking change is unavoidable or the best option when all factors are considered, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen  opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.
```
Comment (jkotas) [inline docs/project/breaking-change-process.md:3]:
```suggestion
Breaking changes make it difficult for people to adopt new versions of .NET. We strive to avoid making them. In the rare case when a breaking change is unavoidable or the best option when all factors are considered, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.
```

**Review by jkotas — APPROVED** [review]

File: docs/project/breaking-change-process.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+   * Engage with people commenting on the issue, with the goal of deriving the most feedback on your proposed breaking change
+   * This may involve significant explanation on your part. A well-written design doc can ease the burden here.
+3. Mark associated PRs with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label, and link to your breaking change issue.
+4. Once the PR is merged, create a [docs issue](https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md).
+   * Clarify which .NET preview the break will be first released in.
+5. Breaking change issues can be closed at any time after the PR is merged. Best practice is waiting until the change has been released in a public preview.
+
+Notes:
+
+* [Guidance for when to use AppContext switched](https://github.com/dotnet/runtime/issues/8655#issuecomment-415103199)
```
Comment (danmosemsft) [inline docs/project/breaking-change-process.md:33]:
```suggestion
* [Guidance for when to use AppContext switches](https://github.com/dotnet/runtime/issues/8655#issuecomment-415103199)
```

**Comment by danmosemsft** [comment]
@PriyaPurkayastha should sign off.

File: docs/project/breaking-change-process.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+You should start this process as soon as possible, preferably before any code is written, in the design phase. This isn't always possible, since the change in behavior might not be discovered until you're almost done building a feature. In the case of a significant breaking change, you should provide reviewers with a design document so that they can understand the broader scope, goals and constraints of your change.
+
+There are people on the team that can help you work through a breaking change, if you want to engage with them before starting this more formal process. They are part of the [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) . Feel free to @mention them on issues to start a dialogue.
+
+## Process
+
+1. Create or link to an issue that describes the breaking change, including the following information:
+   * Mark with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label
+   * Goals and motivation for the change
+   * Pre-change behavior
```
Comment (danmosemsft) [inline docs/project/breaking-change-process.md:16]: Is it interesting to note what versions it affects? Eg., this API was added in .NET Core and the break does not affect folks porting code from .NET Framework.

**Review by danmoseley — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: docs/project/breaking-change-process.md (hunk 12 above)
Comment (richlander) [reply docs/project/breaking-change-process.md:16]: Good call. I added a note on that. This allowed me to bring in relevant context that I'd forgotten about (.NET Framework compatibility mode).

File: docs/project/breaking-change-process.md (hunk 2 above)
Comment (richlander) [reply docs/project/breaking-change-process.md:11]: That one is easy. You should create an issue when you start wanting feedback (not solely on compatibility).

**Review by PriyaPurkayastha — APPROVED** [review]
Looks good to me. Thanks.

**Comment by jkotas** [comment]
Thank you!
